### ISSUE TITLE
Warning C4251 during DLL compilition due to missing IMGUI_API

### ISSUE THREAD

**Issue by ZahlGraf** [ocornut/imgui#345]
Hi

When compiling IMGUI as DLL I get several warnings C4251 like:
```
warning C4251  class "ImVector<ImDrawIdx>" needs to have dll-interface to be used by clients of  struct "ImGuiTextEditState"
```

The background of this warning is, that the struct ImGuiTextEditState is exported with IMGUI_API and it uses the ImVector class that is not exported with IMGUI_API. However when a class is exported and uses other classes/structs as member or base, the other classes and structs must be exported too, otherwise link-errors could occur when using the struct ImGuiTextEditState in the application.

The warning appears for the following classes/structs:
ImVec2 (imgui.h)
ImVec4 (imgui.h)
ImVector (imgui.h)
ImGuiDrawContext (imgui_internal.h)

I had a deeper look at these classes/structs and it seems that all of those are pure inline classes. This means there is not a single method that is not inline. In this case a link-error would not occur, since all necessary information for linking is available.
However my Visual Studio 2015 compiler seems just to check if all classes are exported correctly, independent of if this is really necessary for linking. 
I added IMGUI_API to those classes/structs and was able to compile a DLL without any other warnings.

Best regards,
André

**Comment by ocornut** [comment]
Thanks, will look into it. I would like to make sure that those DLL interfaces don't have a performance effect on both sides of the fence.

**Comment by ZahlGraf** [comment]
Thanks for looking into it.

**Comment by ocornut** [comment]
I looked at this, one problem is that specifying IMGUI_API spreading to template enforce instantiating template specialization within the DLL and application seems to want to link to it there. It's too risky - can't afford letting those inline functions be not-inlined in a DLL context. 

So I have silenced the warning (NB: there are other legit official reasons to silence this warning).

It won't be a problem as long as you use the same version of imgui on both side .lib/.dll anyway.
